sdk/iot/hub/samples/src/paho_iot_hub_c2d_example.c

```diff
@@ -264,40 +264,53 @@ int main()
     printf("Failed to create MQTT client, return code %d\n", rc);
     return rc;
   }
-
-  // Set the callback for incoming MQTT messages
-  if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
-      != MQTTCLIENT_SUCCESS)
-  {
-    printf("Failed to set MQTT callbacks, return code %d\n", rc);
-    return rc;
-  }
-
-  // Connect to IoT Hub
-  if ((rc = connect_device()) != 0)
-  {
-    return rc;
-  }
-
-  // Subscribe to the C2D topic to receive cloud-to-device messages
-  if ((rc = subscribe()) != 0)
-  {
-    return rc;
-  }
-
-  printf("Subscribed to topics.\n");
-
-  // Wait for any incoming C2D messages
-  printf("Waiting for activity. [Press ENTER to abort]\n");
-  (void)getchar();
-
-  if ((rc = MQTTClient_disconnect(mqtt_client, TIMEOUT_MQTT_DISCONNECT_MS)) != MQTTCLIENT_SUCCESS)
+  else
   {
-    printf("Failed to disconnect MQTT client, return code %d\n", rc);
-    return rc;
+    // Set the callback for incoming MQTT messages
+    if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
+        != MQTTCLIENT_SUCCESS)
+    {
+      //Failed to set the callback - move to destroy mqtt client and return
+      printf("Failed to set MQTT callbacks, return code %d\n", rc);
+    }
+    else
+    {
+      // Connect to IoT Hub
+      if ((rc = connect_device()) != 0)
+      {
+        // Connect failed - move to clean up resources
+      }
+      else
+      {
+        // Subscribe to the C2D topic to receive cloud-to-device messages
+        if ((rc = subscribe()) != 0)
+        {
+          // Failed to subscribe - move to disconnect and clean up resources
+        }
+        else
```

Comment:
nit: Can we restructure it to avoid the deep if/else nesting, but rather have it as a sequence of `if(fail condition) print and gracefully fail` stataements?

Reply:
So something like:

```c
  // Connect to IoT Hub
  if ((rc = connect_device()) != 0)
  {
    return rc;
  }

  // Subscribe to the necessary twin topics to receive twin updates and responses
  if ((rc = subscribe()) != 0)
  {
    return rc;
  }
```

becomes

```c
  // Connect to IoT Hub
  if ((rc = connect_device()) != 0)
  {
    MQTTClient_destroy(&mqtt_client);
    return rc;
  }

  // Subscribe to the necessary twin topics to receive twin updates and responses
  if ((rc = subscribe()) != 0)
  {
    MQTTClient_disconnect(mqtt_client, TIMEOUT_MQTT_DISCONNECT_MS);
    MQTTClient_destroy(&mqtt_client);
    return rc;
  }
```

It's a tradeoff...Cascading if/else or repeated destruction code :/

Reply:
Yep. You are right. Either is fine :)
Would using a macros help or is each condition and its block somewhat unique?

Reply:
or....dare I say
.
.
.
.
.
.
.
`goto`

Reply:
But in seriousness they're each a little different. One assumes that the client is initialized and the other initialized and connected

Reply:
I like gotos, so I would be all for using them :)

Reply:
Hmmm. I would consider it safer not to have gotos... 
Some can use it well. Some can cause havoc for using them. Better be safe.

Reply:
or

`    // Set the callback for incoming MQTT messages
    if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
        != MQTTCLIENT_SUCCESS)
    {
      //Failed to set the callback - move to destroy mqtt client and return
      printf("Failed to set MQTT callbacks, return code %d\n", rc);
    }

    if (rc == 0)
    {
      // Connect to IoT Hub
      if ((rc = connect_device()) != 0)
      {
        // Connect failed - move to clean up resources
      }
    }

    if (rc == 0)
    {
        // Subscribe to the C2D topic to receive cloud-to-device messages
        if ((rc = subscribe()) != 0)
        {
          // Failed to subscribe - move to disconnect and clean up resources
        }
    }

    if (rc == 0)
    {
      printf("Subscribed to topics.\n");

      // Connected and subscribed: Wait for any incoming C2D messages
      printf("Waiting for activity. [Press ENTER to abort]\n");
      (void)getchar();
    }`

Reply:
The only problem here I see is that we still need to differentiate between the case in which we need to disconnect and destroy and the case that we just need to destroy.